src/librustc_mir/hair/pattern/_match.rs

```diff
@@ -769,11 +966,42 @@ pub fn is_useful<'p, 'a: 'p, 'tcx: 'a>(cx: &mut MatchCheckCtxt<'a, 'tcx>,
                             witness
                         }).collect()
                     } else {
-                        pats.into_iter().flat_map(|witness| {
-                            missing_ctors.iter().map(move |ctor| {
-                                witness.clone().push_wild_constructor(cx, ctor, pcx.ty)
-                            })
-                        }).collect()
+                        if value_constructors {
+                            // If we've been trying to exhaustively match
+                            // over the domain of values for a type,
+                            // then we can provide better diagnostics
+                            // regarding which values were missing.
+                            missing_ctors.into_iter().map(|ctor| {
+                                match ctor {
+                                    // A constant range of length 1 is simply
+                                    // a constant value.
+                                    ConstantRange(lo, hi, _) if lo == hi => {
```

Comment:
You need to check whether it's inclusive. Also, it might be helpful to never construct `ConstantRange` or even a `PatternKind::Range` that corresponds to exactly one element (including when lowering HIR patterns).

Reply:
It's guaranteed to be inclusive by the code that generates them. But maybe it'd be clearer to move the range -> constant conversion earlier.